syncer/pipe.go

```diff
@@ -0,0 +1,216 @@
+// Pipe is the littlest process unit in syncer
+type Pipe interface {
+	// Name is the pipe's name
+	Name() string
+
+	// Init initializes the pipe, iFunc also do some initialize job.
+	Init(cfg *config.SubTaskConfig, iFunc func() error) error
+
+	// Update updates the pipe's config
+	Update(*config.SubTaskConfig) error
+
+	// Input receives data
+	Input() chan *PipeData
+
+	// Output exports data already be processed
+	Output() chan *PipeData
+
+	// Run starts the pipe's process
+	Run()
+
+	// Close closes the pipe
+	Close()
+
+	// Wait waits all the data in the pipe is processed
+	Wait()
+
+	// SetErrorChan set the error channel, send error to this channel when meet error
+	SetErrorChan(chan error)
+
+	// SetResolveFunc set the resolveFunc, which used when data is resolved, used in the last pipe, or data is skiped
+	SetResolveFunc(func())
+}
+
+// PipeData is the data processed in pipe
+type PipeData struct {
+	binlogEvent *replication.Event
+
+	tp             opType
+	sourceSchema   string
+	sourceTable    string
+	targetSchema   string
+	targetTable    string
+	sqls           []string
+	args           [][]interface{}
+	keys           [][]string
+	retry          bool
+	lastPos        mysql.Position
+	currentPos     mysql.Position // exactly binlog position of current SQL
+	currentGTIDSet gtid.Set
+	ddls           []string
+	traceID        string
+	traceGID       string
+
+	ddlExecItem *DDLExecItem
+
+	// may have other data needed
+	additionalData interface{}
+}
+
+// Pipeline contains many pipes, and resolve data by every pipe.
+type Pipeline struct {
+	pipes []Pipe
+
+	closeCh  chan interface{}
+	isClosed sync2.AtomicBool
+
+	errCh chan error
+
+	wg     sync.WaitGroup
+	dataWg sync.WaitGroup
+}
+
+// NewPipeline returns a new pipeline
+func NewPipeline() *Pipeline {
+	return &Pipeline{
+		pipes: make([]Pipe, 0, 5),
+	}
+}
+
+// AddPipe adds a pipe to this pipeline
+func (p *Pipeline) AddPipe(pipe Pipe) {
+	if p.pipes == nil {
+		p.pipes = []Pipe{pipe}
+		return
+	}
+
+	p.pipes = append(p.pipes, pipe)
+}
+
+// Input receives data
+func (p *Pipeline) Input(data *PipeData) error {
+	if len(p.pipes) == 0 {
+		return errNoPipes
+	}
+
+	if p.isClosed.Get() {
+		return errPipelineClose
+	}
+
+	select {
+	case p.pipes[0].Input() <- data:
+		p.dataWg.Add(1)
+		return nil
+	case <-p.closeCh:
+		return errPipelineCtxDone
+	}
+}
+
+// Flush sends a PipeData with flush type, all pipes should wait all received data is processed
+func (p *Pipeline) Flush() {
+	p.Input(&PipeData{tp: flush})
+	p.Wait()
+}
+
+// Wait waits all pipes have no data to process
+func (p *Pipeline) Wait() {
+	p.dataWg.Wait()
+
+	for _, pipe := range p.pipes {
+		pipe.Wait()
+	}
+}
+
+// Start starts the pipeline's process, run all pipes
+func (p *Pipeline) Start() {
+	log.Info("start pipeline")
+
+	p.closeCh = make(chan interface{})
+	p.errCh = make(chan error, 10)
+
+	for _, pipe := range p.pipes {
+		pipe.SetErrorChan(p.errCh)
+		pipe.SetResolveFunc(func() {
+			p.dataWg.Done()
```

Comment:
when do we call this resolveFunc?

Reply:
had a comment for func `SetResolveFunc`: `SetResolveFunc set the resolveFunc, which used when data is resolved, used in the last pipe, or data is skiped`